File: muxdb-js/src/evaluator/runner.ts
import { Router } from '../router';
import { TelemetryCollector } from '../telemetry';
import { Balancer } from '../balancer';
import { BanditSolver } from '../ml/bandit';
import { ClusterChangeEvent } from './scenario';

export interface SimulationResult {
  policy: string;
  totalQueries: number;
  totalErrors: number;
  avgLatency: number;
  p99Latency: number;
  totalKeysMigrated: number;
  averageImbalanceRatio: number;
  stepMetrics: {
    step: number;
    throughput: number;
    errors: number;
    avgLatency: number;
    imbalanceRatio: number;
  }[];
}

export class SimulationRunner {
  private router: Router;
  private telemetry: TelemetryCollector;
  private balancer: Balancer;
  private banditSolver?: BanditSolver;

  private shardLatencies: Record<string, number> = {};
  private shardCapacities: Record<string, number> = {};
  private shardSlowdowns: Record<string, number> = {};
  private failedShards = new Set<string>();
  private totalKeysMigrated = 0;

  constructor(
    router: Router,
    telemetry: TelemetryCollector,
    balancer: Balancer,
    banditSolver?: BanditSolver,
  ) {
    this.router = router;
    this.telemetry = telemetry;
    this.balancer = balancer;
    this.banditSolver = banditSolver;

    const shards = this.router.shardMap.shards;
    for (const s of shards) {
      this.shardLatencies[s.id] = 5.0;
      this.shardCapacities[s.id] = 100.0;
      this.shardSlowdowns[s.id] = 1.0;
    }
  }

  public applyEvent(event: ClusterChangeEvent): void {
    const shardId = event.targetShardId;
    if (event.actionType === 'ADD_NODE') {
      const currentShards = [...this.router.shardMap.shards];
      if (!currentShards.some((s) => s.id === shardId)) {
        currentShards.push({
          id: shardId,
          backend: 'postgresql',
          host: 'localhost',
          port: 5432,
          database: `db_${shardId}`,
          weight: 1,
          dsn: `postgresql://localhost:5432/db_${shardId}`,
        });
        this.router.shardMap.swap(currentShards);
      }
      this.shardLatencies[shardId] = 5.0;
      this.shardCapacities[shardId] = 100.0;
      this.shardSlowdowns[shardId] = 1.0;
    } else if (event.actionType === 'REMOVE_NODE') {
      const currentShards = this.router.shardMap.shards.filter((s) => s.id !== shardId);
      this.router.shardMap.swap(currentShards);
      this.failedShards.delete(shardId);
    } else if (event.actionType === 'SLOW_NODE') {
      const multiplier = event.parameter !== undefined ? event.parameter : 5.0;
      this.shardSlowdowns[shardId] = multiplier;
    } else if (event.actionType === 'FAIL_NODE') {
      this.failedShards.add(shardId);
    }
  }

  public async run(
    workload: Record<string, any>[][],
    events?: ClusterChangeEvent[],
    policy: 'static' | 'reactive' | 'adaptive' = 'adaptive',
  ): Promise<SimulationResult> {
    const eventsMap: Record<number, ClusterChangeEvent[]> = {};
    if (events) {
      for (const ev of events) {
        if (!eventsMap[ev.step]) {
          eventsMap[ev.step] = [];
        }
        eventsMap[ev.step]!.push(ev);
      }
    }

    const stepMetrics: SimulationResult['stepMetrics'] = [];
    let overallQueries = 0;
    let overallErrors = 0;
    const overallLatencies: number[] = [];

    for (let step = 0; step < workload.length; step++) {
      const stepOps = workload[step] as Record<string, any>[];

      // 1. Apply events
      if (eventsMap[step]) {
        for (const ev of eventsMap[step]!) {
          this.applyEvent(ev);
        }
      }

      const stepLatencies: number[] = [];
      let stepErrors = 0;

      const shardQueriesInStep: Record<string, number> = {};
      for (const sid of Object.keys(this.shardLatencies)) {
        shardQueriesInStep[sid] = 0;
      }

      // 2. Process query ops
      for (const op of stepOps) {
        const key = op.key as string;
        const isWrite = op.op === 'write';

        let targetShardId = '';
        if (policy === 'adaptive' && this.banditSolver) {
          targetShardId = this.banditSolver.selectArm({ type: op.op });
        } else {
          try {
            const shard = this.router.routeKey(key);
            targetShardId = shard.id;
          } catch {
            targetShardId = Object.keys(this.shardLatencies)[0] as string;
          }
        }

        shardQueriesInStep[targetShardId] = (shardQueriesInStep[targetShardId] ?? 0) + 1;

        if (this.failedShards.has(targetShardId)) {
          stepErrors++;
          overallErrors++;
          this.telemetry.recordQuery(targetShardId, key, 1000.0, isWrite);
          continue;
        }

        const base = this.shardLatencies[targetShardId] ?? 5.0;
        const slowdown = this.shardSlowdowns[targetShardId] ?? 1.0;
        const capacity = this.shardCapacities[targetShardId] ?? 100.0;

        const currentQps = shardQueriesInStep[targetShardId] as number;
        let queueingDelay = 0.0;
        if (currentQps > 0) {
          const loadRatio = currentQps / capacity;
          queueingDelay = 10.0 * Math.pow(loadRatio, 2);
        }

        const latency = Math.max(1.0, base * slowdown + queueingDelay);
        this.telemetry.recordQuery(targetShardId, key, latency, isWrite);

        stepLatencies.push(latency);
        overallLatencies.push(latency);
        overallQueries++;

        if (policy === 'adaptive' && this.banditSolver) {
          const reward = 1.0 / (1.0 + latency / 100.0);
          this.banditSolver.updateReward(targetShardId, reward);
        }
      }

      // 3. Dynamic Balancing Decisions
      if (policy !== 'static' && step % 2 === 0) {
        const metrics: Record<string, { read_qps: number; write_qps: number }> = {};
        for (const sid of Object.keys(this.shardLatencies)) {
          metrics[sid] = {
            read_qps: this.telemetry.getQps(sid, false),
            write_qps: this.telemetry.getQps(sid, true),
          };
        }

        const decisions = this.balancer.evaluate(metrics);
        for (const action of decisions) {
          if (action.actionType === 'L2_SPLIT' || action.actionType === 'L3_REPLICA') {
            this.totalKeysMigrated += 50;
            if (action.actionType === 'L2_SPLIT' && Object.keys(this.shardLatencies).length < 8) {
              const newName = `shard_split_${Object.keys(this.shardLatencies).length}`;
              this.applyEvent({ step, actionType: 'ADD_NODE', targetShardId: newName });
            }
          }
        }
      }

      // Compile step stats
      const avgStepLat = stepLatencies.length > 0
        ? stepLatencies.reduce((a, b) => a + b, 0) / stepLatencies.length
        : 0.0;

      const loads = Object.keys(this.shardLatencies)
        .filter((s) => !this.failedShards.has(s))
        .map((s) => shardQueriesInStep[s] as number);
      const avgLoad = loads.length > 0 ? loads.reduce((a, b) => a + b, 0) / loads.length : 1.0;
      const maxLoad = loads.length > 0 ? Math.max(...loads) : 0.0;
      const imbalance = avgLoad > 0 ? maxLoad / avgLoad : 1.0;

      stepMetrics.push({
        step,
        throughput: stepOps.length - stepErrors,
        errors: stepErrors,
        avgLatency: avgStepLat,
        imbalanceRatio: imbalance,
      });
    }

    const avgLat = overallLatencies.length > 0
      ? overallLatencies.reduce((a, b) => a + b, 0) / overallLatencies.length
      : 0.0;
    const sorted = [...overallLatencies].sort((a, b) => a - b);
    const p99Lat = sorted.length > 0 ? (sorted[Math.floor(sorted.length * 0.99)] as number) : 0.0;

    return {
      policy,
      totalQueries: overallQueries,
      totalErrors: overallErrors,
      avgLatency: avgLat,
      p99Latency: p99Lat,
      totalKeysMigrated: this.totalKeysMigrated,
      averageImbalanceRatio: stepMetrics.length > 0
        ? stepMetrics.reduce((sum, m) => sum + m.imbalanceRatio, 0) / stepMetrics.length
        : 1.0,
      stepMetrics,
    };
  }
}

File: muxdb-py/muxdb/evaluator/runner.py
from __future__ import annotations

import math
import random
from typing import List, Dict, Any, Optional

from muxdb.config import MuxConfig, ClusterConfig, ShardConfig
from muxdb.router import Router
from muxdb.telemetry import TelemetryCollector
from muxdb.balancer import Balancer
from muxdb.ml import BanditSolver, DRLTuner
from muxdb.shard_map import ShardInfo
from muxdb.evaluator.scenario import ClusterChangeEvent


class SimulationRunner:
    """
    SimulationRunner drives step-by-step workload replays against a simulated
    MuxDB router and cluster state, tracking observability metrics.
    """

    def __init__(
        self,
        router: Router,
        telemetry: TelemetryCollector,
        balancer: Balancer,
        bandit_solver: Optional[BanditSolver] = None,
        drl_tuner: Optional[DRLTuner] = None,
    ) -> None:
        self.router = router
        self.telemetry = telemetry
        self.balancer = balancer
        self.bandit_solver = bandit_solver
        self.drl_tuner = drl_tuner

        # Simulated cluster health and performance properties
        self.shard_latencies: Dict[str, float] = {}  # Base latency in ms
        self.shard_capacities: Dict[str, float] = {}  # Max QPS capacity
        self.shard_slowdowns: Dict[str, float] = {}  # Latency multipliers
        self.failed_shards: set[str] = set()

        self.knob_state = {"shared_buffers": 256.0, "work_mem": 16.0}
        self.total_keys_migrated = 0
        
        # Initialize defaults for router shards
        for shard in self.router.shard_map.shards:
            self.shard_latencies[shard.id] = 5.0
            self.shard_capacities[shard.id] = 100.0
            self.shard_slowdowns[shard.id] = 1.0

    def apply_event(self, event: ClusterChangeEvent) -> None:
        """Applies cluster hardware state transitions."""
        shard_id = event.target_shard_id
        if event.action_type == "ADD_NODE":
            current_shards = list(self.router.shard_map.shards)
            if not any(s.id == shard_id for s in current_shards):
                current_shards.append(ShardInfo(
                    id=shard_id,
                    backend="postgresql",
                    host="localhost",
                    port=5432,
                    database=f"db_{shard_id}",
                    weight=1,
                    dsn=f"postgresql://localhost:5432/db_{shard_id}",
                ))
                self.router.shard_map.swap(current_shards)
            self.shard_latencies[shard_id] = 5.0
            self.shard_capacities[shard_id] = 100.0
            self.shard_slowdowns[shard_id] = 1.0
        elif event.action_type == "REMOVE_NODE":
            current_shards = [s for s in self.router.shard_map.shards if s.id != shard_id]
            self.router.shard_map.swap(current_shards)
            self.failed_shards.discard(shard_id)
        elif event.action_type == "SLOW_NODE":
            multiplier = event.parameter if event.parameter is not None else 5.0
            self.shard_slowdowns[shard_id] = multiplier
        elif event.action_type == "FAIL_NODE":
            self.failed_shards.add(shard_id)

    def run(
        self,
        workload: List[List[Dict[str, Any]]],
        events: Optional[List[ClusterChangeEvent]] = None,
        policy: str = "adaptive",  # "static", "reactive", "adaptive"
    ) -> Dict[str, Any]:
        """
        Runs the simulation step-by-step.
        Returns:
            Dict: Compiled metrics summary.
        """
        events_map: Dict[int, List[ClusterChangeEvent]] = {}
        if events:
            for event in events:
                events_map.setdefault(event.step, []).append(event)

        step_metrics: List[Dict[str, Any]] = []
        overall_queries = 0
        overall_errors = 0
        overall_latencies = []

        for step, step_ops in enumerate(workload):
            # 1. Apply events scheduled for this step
            if step in events_map:
                for event in events_map[step]:
                    self.apply_event(event)

            # Reset step count
            step_latencies = []
            step_errors = 0
            
            # Temporary counters to calculate queueing delays per shard in this step
            shard_queries_in_step: Dict[str, int] = {s: 0 for s in self.shard_latencies}

            # 2. Route and execute queries
            for op in step_ops:
                key = op["key"]
                is_write = op["op"] == "write"
                
                # Determine target shard using router (static/reactive) or bandit (adaptive)
                if policy == "adaptive" and self.bandit_solver:
                    target_shard_id = self.bandit_solver.select_arm({"type": op["op"]})
                else:
                    try:
                        target_shard = self.router.route_key(key)
                        target_shard_id = target_shard.id
                    except Exception:
                        target_shard_id = list(self.shard_latencies.keys())[0]

                shard_queries_in_step[target_shard_id] = shard_queries_in_step.get(target_shard_id, 0) + 1
                
                # Handle failed shard simulating error
                if target_shard_id in self.failed_shards:
                    step_errors += 1
                    overall_errors += 1
                    # Record query failure in telemetry
                    self.telemetry.record_query(target_shard_id, key, 1000.0, is_write)
                    continue

                # Simulate execution latency: base_latency * slowdown_multiplier + queueing_delay
                base = self.shard_latencies.get(target_shard_id, 5.0)
                slowdown = self.shard_slowdowns.get(target_shard_id, 1.0)
                capacity = self.shard_capacities.get(target_shard_id, 100.0)
                
                # Estimate current QPS load on this shard
                current_qps = shard_queries_in_step[target_shard_id]
                queueing_delay = 0.0
                if current_qps > 0:
                    # Simple queue model: delay spikes when load approaches capacity
                    load_ratio = current_qps / capacity
                    queueing_delay = 10.0 * (load_ratio ** 2)

                # DRL knob adjustment effect (e.g. higher work_mem lowers latency slightly)
                knob_benefit = (self.knob_state["work_mem"] - 16.0) / 10.0
                latency = max(1.0, base * slowdown + queueing_delay - knob_benefit)
                
                # Record to telemetry
                self.telemetry.record_query(target_shard_id, key, latency, is_write)
                
                step_latencies.append(latency)
                overall_latencies.append(latency)
                overall_queries += 1

                # Update bandit reward if policy is adaptive
                if policy == "adaptive" and self.bandit_solver:
                    # Reward function: inverse of latency, normalized
                    reward = 1.0 / (1.0 + (latency / 100.0))
                    self.bandit_solver.update_reward(target_shard_id, reward)

            # 3. Dynamic Balancing Decisions
            if policy != "static" and step % 2 == 0:
                # Gather metrics from telemetry
                metrics: Dict[str, Dict[str, float]] = {}
                for shard_id in self.shard_latencies:
                    metrics[shard_id] = {
                        "read_qps": self.telemetry.get_qps(shard_id, is_write=False),
                        "write_qps": self.telemetry.get_qps(shard_id, is_write=True),
                    }
                
                decisions = self.balancer.evaluate(metrics)
                for action in decisions:
                    if action.action_type in ("L2_SPLIT", "L3_REPLICA"):
                        # Simulate database rebalance data movement
                        self.total_keys_migrated += 50
                        # Balance out: add a virtual balancer node or adjust routing rules
                        if action.action_type == "L2_SPLIT" and len(self.shard_latencies) < 8:
                            new_name = f"shard_split_{len(self.shard_latencies)}"
                            # Register split node
                            self.apply_event(ClusterChangeEvent(step, "ADD_NODE", new_name))

            # 4. Auto-tune database knobs using DRL
            if policy == "adaptive" and self.drl_tuner and step_latencies:
                avg_lat = sum(step_latencies) / len(step_latencies)
                throughput = len(step_latencies)
                
                # Reward is high when throughput is high and latency is low
                reward = (throughput / 100.0) - (avg_lat / 50.0)
                
                new_knobs, loss = self.drl_tuner.update(
                    latency_ms=avg_lat,
                    throughput_qps=float(throughput),
                    current_knobs=self.knob_state,
                    reward=reward,
                )
                self.knob_state = new_knobs

            # Compile step stats
            avg_step_lat = sum(step_latencies) / len(step_latencies) if step_latencies else 0.0
            
            # Calculate imbalance ratio (max load / average load)
            loads = [shard_queries_in_step[s] for s in self.shard_latencies if s not in self.failed_shards]
            avg_load = sum(loads) / len(loads) if loads else 1.0
            max_load = max(loads) if loads else 0.0
            imbalance = max_load / avg_load if avg_load > 0 else 1.0

            step_metrics.append({
                "step": step,
                "throughput": len(step_ops) - step_errors,
                "errors": step_errors,
                "avg_latency": avg_step_lat,
                "imbalance_ratio": imbalance,
            })

        # Compile final stats summary
        avg_lat = sum(overall_latencies) / len(overall_latencies) if overall_latencies else 0.0
        p99_lat = sorted(overall_latencies)[int(len(overall_latencies) * 0.99)] if overall_latencies else 0.0

        return {
            "policy": policy,
            "total_queries": overall_queries,
            "total_errors": overall_errors,
            "avg_latency": avg_lat,
            "p99_latency": p99_lat,
            "total_keys_migrated": self.total_keys_migrated,
            "average_imbalance_ratio": sum(m["imbalance_ratio"] for m in step_metrics) / len(step_metrics) if step_metrics else 1.0,
            "step_metrics": step_metrics,
        }
